### PR TITLE
Update operation docs

### PR DESCRIPTION
**Pull request by mandiwise**
Addresses #250, #584, #728, #754, and #992

## Description

This PR contains updated operations-related content in the Learn docs. Key changes include:

- Expanded content on GraphQL query operations
- A separate "Mutations" page with additional examples for updating and deleting data
- A new "Subscriptions" page
- Updates to the Star Wars schema to support the new mutation examples
- Operation-related content now comes after the Schemas and Types page in the left nav
- Standardization of capitalization and naming for named types

@benjie @jorydotcom

### CODE REVIEW

**Review by Urigo — APPROVED** [review]
so excited for this!

File: src/pages/learn/mutations.mdx
Comment (saihaj) [inline src/pages/learn/mutations.mdx:116]: I think we should clarify that in GraphQL we send a `POST` request to delete

File: src/pages/learn/subscriptions.mdx
Comment (saihaj) [inline src/pages/learn/subscriptions.mdx:1]: For subscriptions I think we should mention that the community maintained specs [SSE](https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md) or [WS](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) which is what most clients & server implementation are build on

File: src/pages/learn/queries.mdx
Comment (saihaj) [inline src/pages/learn/queries.mdx:1]: since we are talking servers, I wonder if we should also mention that there is a spec for GraphQL over HTTP or just link to this doc https://graphql.org/learn/serving-over-http/ as part of next steps for production grade graphql

File: src/pages/learn/mutations.mdx
Comment (mandiwise) [reply src/pages/learn/mutations.mdx:116]: I have an updated "Serving over HTTP" page almost ready to go where I go into this in detail based on the draft GraphQL over HTTP spec (including when `GET` may be used too). Does it make sense to duplicate that here? My goal was draw connections with REST patterns the reader would likely be familiar with while explaining how different kinds of mutations work without overloading with too much information. It can be tricky to find the right balance though! Let me know what you think.

File: src/pages/learn/queries.mdx
Comment (mandiwise) [reply src/pages/learn/queries.mdx:1]: Added a link to the recap section.

File: src/pages/learn/subscriptions.mdx
Comment (mandiwise) [reply src/pages/learn/subscriptions.mdx:1]: Good call out. Added links to these now.

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
````suggestion
At its simplest, GraphQL is about asking for specific [fields](https://spec.graphql.org/draft/#sec-Language.Fields) on objects. Let's start by looking at the `hero` field that's defined on the `Query` type in the schema:

```graphql
type Query {
  hero: Character
}
```

We can see what result we get when we query it:
````

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
When creating a GraphQL _document_ we always start with a [root operation type](/learn/schema/#the-query-mutation-and-subscription-types) (the `Query` Object type for this example) because it serves as an entry point to the API. From there we must specify the _selection set_ of fields we are interested in, all the way down to their leaf values which will be Scalar or Enum types. The field `name` returns a `String` type, in this case the name of the main hero of Star Wars, `"R2-D2"`.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx:32]: Would be nice to have something like: "We'll see more on why this is so important in [link to docs]" because it's not just specifically to gain the performance and efficiency advantages of avoiding over-/under-fetching as many think, but also relates to versioning (including deprecation and field removal), error handling (preventing selection sets blowing up unnecessarily), symmetry (receiving exactly what you ask for), locality (each component declaring its own dependencies and being able to be edited without breaking other components' data needs), and shaping the response in a client-focussed way.

But no need to address this in this PR specifically.

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]: I don't think this first edit is desired; we're talking about the field `name: String`, so "more deeply nested Object type" doesn't seem to make sense here since it's not returning an Object type (nested or otherwise).

```suggestion
In the previous example, we just asked for the name of our hero which returned a `String`, but fields can also return Object types (and lists thereof). In that case, you can make a _sub-selection_ of fields for that Object type:
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
You can even pass arguments into fields that output Scalar types; one use case for this would be to implement data transformations once on the server, instead of on every client separately:
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
In the examples above we have been using a shorthand syntax where we omit the `query` keyword before the operation's selection set. In addition to specifying the _operation type_ explicitly, we can also add a unique _operation name_, which is useful in production apps because it makes debugging and tracing easier.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
The operation type is either `query`, `mutation`, or `subscription` and describes what type of operation you intend to do. This keyword is required unless you're using the shorthand syntax for queries (it is always required for mutations and subscriptions). Additionally, if you wish to provide a name for your operation, then you must specify the operation type as well.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
The operation name is an explicit name that you assign to your operation; you should pick a meaningful name. It is required when sending multiple operations in one document, but even if you're only sending one operation it's encouraged because operation names are helpful for debugging and server-side logging. When something goes wrong (you see errors either in your network logs or in the logs of your GraphQL server) it is easier to identify a query in your codebase by name instead of trying to decipher the contents.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
Let's say we have a relatively complicated page in our app, which lets us look at two heroes side by side, along with their friends. You can imagine that such a query could quickly get complicated because we would need to repeat the fields at least once—one for each side of the comparison.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
You can see how the above query would be pretty repetitive if we weren't able to use fragments. The concept of fragments is frequently used to split complicated application data requirements into smaller chunks, especially when you need to combine many UI components with different fragments into one initial data fetch.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]: "A few" is typically interpreted as "3 or more"; but I think other than `__typename` there's only `__schema` and `__type`. Let's just list them:

```suggestion
All field names beginning with two underscores (`__`) are reserved by GraphQL. In addition to `__typename`, GraphQL services provide the `__schema` and `__type` meta-fields which expose the [introspection](/learn/introspection) system.
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
- Operations should be given unique names, which make requests more expressive and help with debugging
```

File: src/pages/learn/queries.mdx
Comment (benjie) [inline src/pages/learn/queries.mdx]:
```suggestion
Now that we understand the ins and outs of how to read data from a GraphQL server with query operations, it's time to learn how to change data and trigger side effects using [mutations](/learn/mutations/).
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]:
```suggestion
In REST, any request might cause some side-effects on the server, but by convention, it's suggested that one doesn't use `GET` requests to modify data. GraphQL is similar—technically any field resolver could be implemented to cause a data write—but the [GraphQL specification states](https://spec.graphql.org/draft/#sel-GANRNDAB6DBmMn6D):

> the resolution of fields other than top-level mutation fields must always be side effect-free and idempotent

Thus only the top-level fields in mutation operations are allowed to cause side effects; to do otherwise would make your schema GraphQL non-compliant.
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]: I am very uncomfortable with this positioning; GraphQL isn't just a CRUD system and in general it's better to use semantically named mutations for specific user actions rather than trying to fit every mutation into the CRUD box. Some mutations such as `processPayment`, `resetPassword`, `sendEmailVerification`, `inviteUser`, `markAllNotificationsAsRead`, `redeemGiftCard` and so on (arguably) don't even fit into the CRUD box. (This isn't to say that CRUD should never be used, but more that a semantically named action should be exposed preferentially where it makes sense to do so.)

From GraphQL's point of view there are no different types of mutation - there is no difference between a create, update, delete or arbitrary user action mutation - so I'm concerned that highlighting CRUD mutations in this way sets the reader (especially someone who does not read the page fully) up for misconceptions.

**Review by benjie — CHANGES_REQUESTED** [review]
I love the improvements to the queries page (and include a number of minor suggested edits), but I have concerns over the mutations page. I have not yet read the subscriptions page.

File: src/pages/learn/mutations.mdx
Comment (mandiwise) [reply src/pages/learn/mutations.mdx]: I see where you're coming from. I think the challenge here is to figure out how much information to include up-front in the intro for a reader who has some experience with REST and needs to connect those dots with GraphQL (e.g. #992). This distinction is called out in the paragraph that follows the new example for updating a human name, but it's halfway down the page.

Do you think it would be better to generalize it like this?

> On this page, you'll learn how to use mutation operations to modify data on a server using GraphQL.

Comment (mandiwise) [reply src/pages/learn/mutations.mdx]: Do you think the subheads on the page are problematic? We could further deemphasize any kind of 1:1 CRUD implication by renaming them like this:

~Create data~ Add new data
~Update data~ Update existing data
~Delete data~ Remove existing data

Comment (mandiwise) [reply src/pages/learn/mutations.mdx]: One more thought on this is—I think there's a gap in the Best Practices content where "migrating from REST" could be covered in more detail too with a few more detailed examples that are based on a further-enhanced version of the Star Wars schema.

Comment (benjie) [reply src/pages/learn/mutations.mdx]: I'm not sure that our learn docs should necessarily cater primarily to people who already know REST/CRUD - perhaps a separate guide for "how to convert CRUD REST operations to GraphQL" might make sense but I think our learn docs shouldn't start with REST concepts and map them to GraphQL, but instead start with GraphQL best practices and then if necessary expand with REST accommodations/analogies.

I think my preferred way of addressing this would be to start with a non-CRUD mutation ~~that follows best practices (input object, mutation payload)~~ so that that is front and centre, and then maybe have a sub heading about CRUD that has examples of create, update, upsert, delete. I think the original documentation used a create mutation because that's fairly universal, but perhaps pivoting away from that would help learners to do GraphQL the right way from the start?

Trying to think of something within the Star Wars schema that would make sense as a non-CRUD mutation, perhaps `rateFilm` which stores a rating "onto the film" from your perspective, but clearly isn't actually updating the film itself:

```graphql
type Mutation {
  rateFilm(input: RateFilmInput!): RateFilmPayload
}

input RateFilmInput {
  filmId: ID!
  rating: FilmRating! = THUMBS_UP
}

type RateFilmPayload {
  film: Film
}

type Film {
  # ...
  viewerRating: FilmRating
}

enum FilmRating {
  THUMBS_DOWN
  NEUTRAL
  THUMBS_UP
  DOUBLE_THUMBS_UP
}
```

Comment (benjie) [reply src/pages/learn/mutations.mdx]: Actually I think this page should talk about the input object/payload best practices at the end, so maybe simplify this down:

```graphql
type Mutation {
  rateFilm(
    filmId: ID!
    rating: FilmRating! = THUMBS_UP
  ): Film
}

type Film {
  # ...
  viewerRating: FilmRating
}

enum FilmRating {
  THUMBS_DOWN
  NEUTRAL
  THUMBS_UP
  DOUBLE_THUMBS_UP
}
```

Comment (mandiwise) [reply src/pages/learn/mutations.mdx]: I can come up with something but at this point I think I'll wait until #1812 is merged so I rebase the changes made in `swapi-schema.tsx` in that PR first.

Comment (benjie) [reply src/pages/learn/mutations.mdx]: - #1812

is now merged :+1:

Comment (mandiwise) [reply src/pages/learn/mutations.mdx]: There's a new commit with the requested changes included now.

**Comment by benjie** [comment]
Not quite sure what happened here but it went to 136 files changed... I've merged in the latest `source` and resolved the conflicts... That seems to have us back down to 5 files changed.

File: src/pages/learn/subscriptions.mdx
Comment (benjie) [inline src/pages/learn/subscriptions.mdx]: Edited for precision (events don't necessarily have to update existing data; fetch implies client-pull but subscriptions are server-push):

```suggestion
In addition to reading and writing data using stateless query and mutation operations, the GraphQL specification also describes how to receive real-time updates via long-lived requests. On this page, we'll explore how clients can subscribe to details of events on a GraphQL server using subscription operations.
```

File: src/pages/learn/subscriptions.mdx
Comment (benjie) [inline src/pages/learn/subscriptions.mdx]: Let's make it a little clearer that there's no standard implementation of live queries.

```suggestion
Note that subscriptions are different from the concept of _live queries_, a loosely defined feature that some GraphQL implementations may offer. [Read more about the ongoing discussion on a formal specification for live queries here.](https://github.com/graphql/graphql-spec/issues/386)
```

File: src/pages/learn/subscriptions.mdx
Comment (benjie) [inline src/pages/learn/subscriptions.mdx]:
```suggestion
But this document is valid:
```

File: src/pages/learn/subscriptions.mdx
Comment (benjie) [inline src/pages/learn/subscriptions.mdx]:
```suggestion
As with query and mutation operations, the document above contains multiple operations so each operation must be named, and the name of the operation to execute must be specified in the request.
```

File: src/pages/learn/subscriptions.mdx
Comment (benjie) [inline src/pages/learn/subscriptions.mdx]: The text talked about subscriptions relating to "data that changes", but subscriptions are really well suited for events such as "Bob is typing..." or "Updated GraphQL schema available" that may not actually have a representation in the data and thus polling may not even be an option. I've tweaked this wording so that it doesn't exclude the pure-event use case:

```suggestion
Subscription operations are well suited for data that changes often and incrementally, and for clients that need to receive those incremental updates as close to real-time as possible to deliver the expected user experience. For data with less frequent updates, periodic polling, mobile push notifications, or re-fetching queries based on user interaction may be the better solutions for keeping a client UI up to date.
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]: Let's not imply that the input object itself is "created":

```suggestion
Mutation fields can also accept arguments and you might notice that the `review` argument has an input type set to `ReviewInput`. This is known as [Input Object type](/learn/schema/#input-object-types), which allows us to pass in a structured object containing information the mutation can use instead of individual scalar values only.
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]: I don't want to imply that `createReview(...): Review` is the convention people should follow when `createReview(...): CreateReviewPayload` is a much better pattern. I've changed "corresponds" to "relates" since the `CreateReviewPayload` and `Review` both _relate_ to the type that's modified, whereas I'm not sure they both "correspond" to it? (I'm no English scholar though!)

```suggestion
While the `createReview` field could be defined with any valid output type in the schema, it's conventional to specify an output type that relates to whatever is modified during the mutation—in this case, the `Review` type. This can be useful for clients that need to fetch the new state of an object after an update.
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]: Missing word?

```suggestion
You can learn more about how GraphQL provides data for fields on the [Execution page](/learn/execution).
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]: This is already really good, but I've edited it to make it even clearer this is a choice.

```suggestion
This example demonstrates an important distinction from REST. To update a human's properties using a REST API, you would likely send any updated data to a generalized endpoint for that resource using a `PATCH` request. With GraphQL, instead of creating a generic `updateHuman` mutation, you may choose to define more specific mutation fields such as `updateHumanName` that are designed for the task at hand.
```

File: src/pages/learn/mutations.mdx
Comment (benjie) [inline src/pages/learn/mutations.mdx]: This is really great content, but "get the data that they need from" seems very query-focussed (and side-effect free) - can you tweak it so that it better reflects mutations?

File: src/pages/learn/mutations.mdx
Comment (benjie) [reply src/pages/learn/mutations.mdx]: Should this section have it's own header: `## Purpose-driven mutations`

**Review by benjie — CHANGES_REQUESTED** [review]
I like this new structure and this is great content - really really good! I've included some minor nit-picky suggestions below - the only big change I think is that it would be nice to see the purpose-driven mutations content raised to be its own heading so it's easy to link to, and doing so may or may not require some minor reworking of the surrounding text.

File: src/pages/learn/mutations.mdx
Comment (mandiwise) [reply src/pages/learn/mutations.mdx]: Fixed in https://github.com/graphql/graphql.github.io/pull/1802/commits/1f3ae06e30c231b965c8f591ababce6d9c88ae8f.

Comment (benjie) [reply src/pages/learn/mutations.mdx]: Perfect 🙌

**Review by benjie — APPROVED** [review]
Looks great to me!

**Pull request merged**
